vllm/model_executor/models/bamba.py

```diff
@@ -0,0 +1,610 @@
+
+        self.rotary_emb = RotaryEmbedding(
+            head_size=self.head_dim,
+            rotary_dim=config.attn_rotary_emb,
+            max_position_embeddings=max_position_embeddings,
+            base=rope_theta,
+            is_neox_style=True,
+            dtype=torch.get_default_dtype(),  # see impl of get_rope
+        )
+
+        self.qkv_proj = QKVParallelLinear(
+            config.hidden_size,
+            self.head_dim,
+            self.total_num_heads,
+            self.total_num_kv_heads,
+            bias=False,
+            quant_config=quant_config,
+        )
+        self.o_proj = RowParallelLinear(self.total_num_heads * self.head_dim,
+                                        config.hidden_size,
+                                        bias=False,
+                                        quant_config=quant_config)
+
+        self.attn = Attention(
+            self.num_heads,
+            self.head_dim,
+            self.scaling,
+            num_kv_heads=self.num_kv_heads,
+            cache_config=cache_config,
+            prefix=f"{prefix}.attn",
+        )
+
+        self.feed_forward = BambaMLP(config, quant_config=quant_config)
+        self.input_layernorm = RMSNorm(config.hidden_size,
+                                       eps=config.rms_norm_eps)
+        self.pre_ff_layernorm = RMSNorm(config.hidden_size,
+                                        eps=config.rms_norm_eps)
+
+    def self_attention(
+        self,
+        positions: torch.Tensor,
+        hidden_states: torch.Tensor,
+        kv_cache: torch.Tensor,
+        attn_metadata: AttentionMetadata,
+        **kwargs,
+    ) -> torch.Tensor:
+        qkv, _ = self.qkv_proj(hidden_states)
+        q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
+
+        # because the bamba model may potentially handle long sequences,
+        # we should adjust the sin_cos cache if necessary to avoid out of bounds
+        # - first get the max_position
+        max_position = max(
+            getattr(attn_metadata, 'max_prefill_seq_len', 0),
+            getattr(attn_metadata, 'max_decode_seq_len', 0),
+        )
+        if max_position == 0:
+            # if we cannot get the max length from the metadata, then
+            # get it from the positions
+            max_position = positions.max().item()
+
+        # when VLLM_ALLOW_LONG_MAX_MODEL_LEN=1 could potentially cause inputs
+        # longer than max_position_embeddings. We extend the rope cache
+        # to prevent CUDA errors. Be aware that the outputs could be of
+        # lower quality for long sequence lengths.
+        rotary = self.rotary_emb
+        if rotary.max_position_embeddings <= max_position:
+            # we set it to the next power of two that covers it
+            while rotary.max_position_embeddings <= max_position:
+                rotary.max_position_embeddings *= 2
+            rotary.cos_sin_cache = rotary._compute_cos_sin_cache()
```

Comment:
Instead of this, have you considered using `rope_scaling` instead? If you use `get_rope` instead of constructing the RotaryEmbedding directly, I think it should work for `bamba`

See this unit test for an example of how it works:
https://github.com/sasha0552/vllm/blob/d427e5cfda8d2536b81e6021128e71b2dbc281aa/tests/test_config.py#L177

See also:
https://www.reddit.com/r/LocalLLaMA/comments/14mrgpr/dynamically_scaled_rope_further_increases/